anitya/lib/model.py

```diff
@@ -908,3 +919,191 @@ def last_entry(cls, session):
+        username (str): The user's username, as retrieved from third-party authentication.
+        active (bool): Indicates whether the user is active. If false, users will not be
+            able to log in.
+        social_auth (sqlalchemy.orm.dynamic.AppenderQuery): The list of
+            :class:`social_flask_sqlalchemy.models.UserSocialAuth` entries for this user.
+    """
+    __tablename__ = 'users'
+
+    id = sa.Column(GUID, primary_key=True, default=uuid.uuid4)
+    # SMTP says 256 is the maximum length of a path:
+    # https://tools.ietf.org/html/rfc5321#section-4.5.3
+    email = sa.Column(sa.String(256), nullable=False, index=True, unique=True)
+    username = sa.Column(sa.String(256), nullable=False, index=True, unique=True)
+    active = sa.Column(sa.Boolean, default=True)
+
+    @property
+    def admin(self):
+        """
+        Determine if this user is an administrator.
+
+        Returns:
+            bool: True if the user is an administrator.
+        """
+        return six.text_type(self.id) in anitya_config.get('ANITYA_WEB_ADMINS', [])
+
+    @property
+    def is_active(self):
+        """
+        Implement the flask-login interface for determining if the user is active.
+
+        If a user is _not_ active, they are not allowed to log in.
+
+        Returns:
+            bool: True if the user is active.
+        """
+        return self.active
+
+    @property
+    def is_anonymous(self):
+        """
+        Implement the flask-login interface for determining if the user is authenticated.
+
+        flask-login uses an "anonymous user" object if there is no authenticated user. This
+        indicates to flask-login this user is not an anonymous user.
+
+        Returns:
+            bool: False in all cases.
+        """
+        return False
+
+    @property
+    def is_authenticated(self):
+        """
+        Implement the flask-login interface for determining if the user is authenticated.
+
+        In this case, if flask-login has an instance of :class:`User`, then that user has
+        already authenticated via a third-party authentication mechanism.
+
+        Returns:
+            bool: True in all cases.
+        """
+        return True
+
+    def get_id(self):
+        """
+        Implement the flask-login interface for retrieving the user's ID.
+
+        Returns:
+            six.text_type: The Unicode string that uniquely identifies a user.
+        """
+        return six.text_type(self.id)
+
+
+def _api_token_generator(charset=string.ascii_letters + string.digits, length=40):
+    """
+    Generate an API token of a given length using the provided character set.
+
+    Args:
+        charset (str): A string of characters to choose from in the token.
+        length (int): The number of characters to use in the token.
+
+    Returns:
+        str: The API token as a unicode string.
+    """
+    return u''.join(random_choice(charset) for __ in range(length))
```

Comment:
Note the warning in the [random](https://docs.python.org/3.5/library/random.html) docs:

```
Warning The pseudo-random generators of this module should not be used for security purposes.
```

I saw at the top that you use the Python 3.6 [```secrets```](https://docs.python.org/3.6/library/secrets.html) module if available, which does say that it can be used for "generating cryptographically strong random numbers suitable for managing data such as passwords, account authentication, security tokens, and related secrets."

Reply:
Good catch, I added that and meant to come back and find the equivalent non-Python 3.6 API, but clearly didn't. I changed it so that if secrets isn't available, it falls back to [SystemRandom](https://docs.python.org/3.5/library/random.html#random.SystemRandom) which uses ``os.urandom`` rather than the Mersenne Twister for random numbers.

This is in line with the [recommendations of cryptography](https://cryptography.io/en/stable/random-numbers/) and is what Django does as well.